telemetry/client.go

```diff
@@ -10,70 +10,207 @@ import (
 }
 
-func NewClient(logger *zap.Logger, serverURL string, keyUID string, nodeName string) *Client {
+func NewClient(logger *zap.Logger, serverURL string, keyUID string, nodeName string, version string) *Client {
 	return &Client{
-		serverURL:  serverURL,
-		httpClient: &http.Client{Timeout: time.Minute},
-		logger:     logger,
-		keyUID:     keyUID,
-		nodeName:   nodeName,
+		serverURL:          serverURL,
+		httpClient:         &http.Client{Timeout: time.Minute},
+		logger:             logger,
+		keyUID:             keyUID,
+		nodeName:           nodeName,
+		version:            version,
+		receivedMessagesCh: make(chan wakuv2.ReceivedMessages),
+		receivedEnvelopeCh: make(chan *v2protocol.Envelope),
+		sentEnvelopeCh:     make(chan wakuv2.SentEnvelope),
+		telemetryCh:        make(chan TelemetryRequest),
+		nextId:             0,
+		sendPeriod:         10 * time.Second,
+	}
+}
+
+func (c *Client) CollectAndProcessTelemetry() {
+	go func() {
+		for {
+			select {
+			case receivedMessages := <-c.receivedMessagesCh:
+				c.processAndPushTelemetry(receivedMessages)
+			case receivedEnvelope := <-c.receivedEnvelopeCh:
+				c.processAndPushTelemetry(receivedEnvelope)
+			case sentEnvelope := <-c.sentEnvelopeCh:
+				c.processAndPushTelemetry(sentEnvelope)
+			}
+		}
+	}()
+}
+
+func (c *Client) Start() {
+	go c.CollectAndProcessTelemetry()
+	go func() {
+		ticker := time.NewTicker(c.sendPeriod)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				var telemetryRequests []TelemetryRequest
+				collecting := true
+				for collecting {
+					select {
+					case telemetryRequest := <-c.telemetryCh:
+						telemetryRequests = append(telemetryRequests, telemetryRequest)
+					default:
+						collecting = false
+					}
+				}
+				if len(telemetryRequests) > 0 {
+					c.pushTelemetryRequest(telemetryRequests)
+				}
+			}
+		}
+	}()
+}
+
+func (c *Client) processAndPushTelemetry(data interface{}) {
+	var telemetryRequest TelemetryRequest
+	switch v := data.(type) {
+	case wakuv2.ReceivedMessages:
+		telemetryRequest = TelemetryRequest{
+			Id:            c.nextId,
+			TelemetryType: ReceivedMessagesMetric,
+			TelemetryData: c.ProcessReceivedMessages(v),
+		}
+	case *v2protocol.Envelope:
+		telemetryRequest = TelemetryRequest{
+			Id:            c.nextId,
+			TelemetryType: ReceivedEnvelopeMetric,
+			TelemetryData: c.ProcessReceivedEnvelope(v),
+		}
+	case wakuv2.SentEnvelope:
+		telemetryRequest = TelemetryRequest{
+			Id:            c.nextId,
+			TelemetryType: SentEnvelopeMetric,
+			TelemetryData: c.ProcessSentEnvelope(v),
+		}
+	default:
+		c.logger.Error("Unknown telemetry data type")
+		return
 	}
+
+	c.nextId++
+	c.telemetryCh <- telemetryRequest
 }
 
-func (c *Client) PushReceivedMessages(filter transport.Filter, sshMessage *types.Message, messages []*v1protocol.StatusMessage) {
+func (c *Client) pushTelemetryRequest(request []TelemetryRequest) {
+	c.logger.Debug("Pushing telemetry data", zap.Any("request", request))
```

Comment:
Consider not logging this unless it's strictly necessary

Reply:
removed in latest commit